components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/SubscriberRegistrationInterceptor.java

```diff
@@ -55,31 +58,50 @@ public void handleMessage(Message message) {
             return;
         }
 
+        // checking if the subscriber exists in the subscriber cache
+        Cache<String, Subscriber> subscriberCache = Caching.getCacheManager(APIConstants.API_MANAGER_CACHE_MANAGER)
+                .getCache(APIConstants.API_SUBSCRIBER_CACHE);
+        if (subscriberCache.get(username) != null) {
+            return;
+        }
+
+        // check the existence in the database
         String groupId = RestApiUtil.getLoggedInUserGroupId();
         String tenantDomain = RestApiUtil.getLoggedInUserTenantDomain();
         try {
             //takes a consumer object using the user set in thread local carbon context
             APIConsumer apiConsumer = RestApiUtil.getLoggedInUserConsumer();
             Subscriber subscriber = apiConsumer.getSubscriber(username);
             if (subscriber == null) {
-                try {
-                    APIUtil.checkPermission(username, APIConstants.Permissions.API_SUBSCRIBE);
-                } catch (APIManagementException e) {
-                    // When user does not have subscribe permission we will log it and continue flow.
-                    // This happens when user tries to access anonymous apis although he does not have subscribe 
-                    // permission. It should be allowed.
-                    if (logger.isDebugEnabled()) {
-                        logger.debug("User " + username + " does not have subscribe permission", e);
+                synchronized (this) {
```

Comment:
do we need to sync on 'this'? syncing only on the `username+somestring` is not enough?

Reply:
Should be okay. Will fix that.